torch_xla/utils/utils.py

```diff
@@ -8,6 +8,42 @@
 import sys
 import tempfile
 import time
+import json
+from collections import defaultdict
+
+import torch_xla.utils.gcsfs as gcsfs
+
+
+class CheckpointTagger(object):
+
+  def __init__(self, tags=None, refcount=None, remover=None):
+    self.tags = dict() if tags is None else tags
+    self._refcount = defaultdict(int) if refcount is None else refcount
+    self.set_remover(remover)
+
+  def tag(self, name, path):
+    self._refcount[path] += 1
+    old_path = self.tags.get(name)
+    if old_path is not None:
+      self._refcount[old_path] -= 1
+      if self._refcount[old_path] == 0:
+        self._refcount.pop(old_path)
+        self._remover(old_path)
+    self.tags[name] = path
+
+  def set_remover(self, remover):
+    remover = (lambda x: None) if remover is None else remover
+    assert callable(remover)
+    self._remover = remover
+
+  def save(self, filename):
+    dat = json.dumps({'tags': self.tags, 'refcount': self._refcount})
```

Comment:
Do not save refcount.
You can rebuild it.
The reload can just be a serie of .tag() calls.

Reply:
makes sense, done.